sdk/include/opentelemetry/sdk/metrics/meter_context.h

```diff
@@ -91,23 +109,34 @@ class MeterContext
                std::unique_ptr<View> view) noexcept;
 
   /**
-   * Force all active Exporters and Readers to flush any buffered meter data
+   * Adds a meter to the list of configured meters.
+   *
+   * Note: This method is not thread safe, and should ideally be called from main thread in
+   * thread-safe manner.
+   *
+   * @param meter
+   */
+  void AddMeter(std::shared_ptr<Meter> meter);
+
+  /**
+   * Force all active Exporters and Collectors to flush any buffered meter data
    * within the given timeout.
    */
 
   bool ForceFlush(std::chrono::microseconds timeout = (std::chrono::microseconds::max)()) noexcept;
 
   /**
-   * Shutdown the Exporters and Readers associated with this meter provider.
+   * Shutdown the Exporters and Collectors associated with this meter provider.
```

Comment:
Should `Shutdown` and `ForceFlush`, be thead safe?

Reply:
No, Shutdown and Forceflush are only supposed to be called from MeterProvider on the main thread.

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#concurrency-requirements

Reply:
Thanks for correcting me. Need to fix it in that case.

Reply:
Ok looking further into the code, I think we should be good. It is safe to call Shutdown and Foreflush concurrently on MeterProvider- as these methods eventually invoke the corresponding methods for MeterReader and MeterExporter, and they are ensured to be thread-safe.

Reply:
Sounds good. FYI in OpenTelemetry .NET we have something similar:

The public `Shutdown` is thread safe:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/bab645fe3dadc46235eef1e29c9959301b9c6e90/src/OpenTelemetry/Metrics/MetricReader.cs#L155-L159

The protected `OnShutdown` is not thread safe, and the SDK guarantees that `OnShutdown` is called in a synchronized manner:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/bab645fe3dadc46235eef1e29c9959301b9c6e90/src/OpenTelemetry/Metrics/MetricReader.cs#L307-L312

Reply:
Thanks for sharing. We do something similar too in this PR. `MetricReader::Shutdown()` is thread-safe, and pure virtual method `MetricReader::OnShutdown()` to be implemented by custom readers is not supposed to be thread-safe.